### DISCUSSION TITLE
NetNamedPipeBinding - please release it

### DISCUSSION THREAD

**Discussion by cbenard** [CoreWCF/CoreWCF#1412] [General]
Hi @mconnew. `NetNamedPipeBinding` is really a blocker for us. We have some Win10 users and can't use Unix Socket binding and we are trying to update a very large line-of-business app to .NET 8. 

We really need `NetNamedPipeBinding`. Here's a timeline:

* `1.4.0-preview1` of it was released in April 2023.
* `1.5.0-preview1` was released in September 2023.
* `1.5.1` and `1.5.2` have been released since then, which have included the later-added `UnixDomainSocketBinding`, but have not included the seemingly working `NetNamedPipeBinding`.

Having it in a `-release1` NuGet prerelease prefix stops us from using it organizationally. Could you see your way to "blessing" it as release, as the `UnixDomainSocketBinding` has been?

**Comment by mconnew** [comment]
I never intended to take this long to circling back to finish off NetNamedPipeBinding. There are a few tasks needed to finish the work, which I intended to get to relatively quickly, but other priorities have pushed it down my task list multiple times.  

The code is release quality, it's simply missing a few small things, and has a behavior difference which can range from annoying to a blocker depending on how your existing code is written. It doesn't cleanly close the connections on service shutdown, so the clients see an aborted connection instead of a closed connection. It's missing WSDL support. And there's a problem where it requires configuring kestrel to listen on a benign port.  

It's been punted down the road longer than is really helpful, and lots of people are asking for it to be out of preview. If there's clear messaging that these shortcomings have to be lived with until I can get to them in a later release so wouldn't be eligible to be asked to be fixed as part of opening a support case with Microsoft, would that be an okay situation if I dropped the preview label and made it a release?

**Reply by cbenard** [reply]
We don't need WSDL at all, since we're using code-first interface sharing, so that wouldn't affect me. I can't speak for others.

As far as "not cleanly closing" vs "aborted connection", if the channel still becomes `Faulted` status, our "re-initiate the connection" code would work as-is. I'm not sure if it would get set to `Faulted` or not based on the description above.

If the `Faulted` status is set, I'd personally be fine with those limitations, but perhaps organizationally, I can say "the guy that runs the project said these are the limitations" we could use the `-release1`.

My fear, if you release it with the connection limitation (I don't care about WSDL here at all) is that perhaps that removes it from the visible roadmap and it's not addressed.

I'd say "keep it on the roadmap and keep it in pre-release and hopefully complete some time soon-ish" if it has issues that need to be addressed to be feature-complete and I can see if I can get authorization to use it with the pre-release NuGet.

I do need an answer on the `Faulted` part though. Our actual implementation catches `FaultException` and `CommunicationObjectFaultedException` on calls and then tears down/sets up again.

**Reply by mconnew** [reply]
I can't remember the exact exception the client sees. A clean shutdown has a close handshake that the service initiates which is basically a single byte sent (turns into more when encrypted over NegotiateStream due to encryption padding etc, but the consuming code sees a single byte), then the client sends a single byte back, then they both consider the channel closed. That doesn't close the actual connection as WCF/CoreWCF enables connection reuse (think of it like HTTP connection pooling, but at the channel layer and not request layer). Here's where there's another subtlety, the underlying actual Windows named pipe supports sending zero byte payloads. This implementation is built on top of the runtimes named pipe implementation instead of doing all the work ourselves using P/Invoke. The .NET runtime implementation sees the payload we're trying to send is zero bytes and no-op's the call, so the zero byte payload never gets sent. Normally WCF will receive the zero byte payload and take it as the actual connection is closing (same way a clean close/shutdown happens with sockets). That's not possible so WCF will presume the connection is going to be reused and stick it in the idle pool. When it goes to use it for a new connection, it will get an exception about the connection handle being invalid, and create a new connection instead. Your app code won't see this, it gets swallowed as it's normal if the service process was restarted abruptly.  

The easiest way to see the behavior different is to run the service, connect with a client and make a call. Then stop the service but leave the process running and see what exception the client sees. Repeat again, but this time exiting the service process. I think you'll see CommunicationObjectFaultedException in both cases. A FaultException should be what's thrown when your service call throws an exception when executing your service call and I don't think there's a communications failure which would trigger it. All WCF exceptions derive from CommunicationException except TimeoutException. You can do a try/catch with multiple catch blocks and can cover all communication exceptions easily like this:
```c#
try
{
    channel.DoSomething();
}
catch (FaultException fe)
{
    // handle a FaultException from the service
}
catch (CommunicationException ce)
{
    // This includes CommunicationObjectFaultedException and anything else that might get thrown
}
catch (TimeoutException te)
{
    // Timeouts throw the BCL TimeoutException so don't derive from CE
}
```
This should cover you regardless of what exception is thrown, and should keep your application level exceptions handled independently.

**Reply by cbenard** [reply]
Regarding the zero-byte things, we would never have a zero-byte payload, due to the DataContractSerializer and envelope/etc for request/response. Are you saying something underneath might generate a zero-byte request and get no-opped, resulting in an error?

If that's the case, this could never be addressed since it sits on top of the runtime and not P/Invoke, right?

I'm just trying to determine the risk exposure here of unexpected errors at runtime at client sites that we might not see in dev or QA.

**Reply by mconnew** [reply]
The zero byte payload is an internal thing to NetNamedPipe. There is no close convention on a named pipe like there is on a socket. If one end closes the handle, the other end gets a rude invalid handle error. The zero bytes is just some NetNamedPipe book keeping to tell the other end the connection handle is being closed.

**Reply by mconnew** [reply]
This problem can be solved as the runtime class does expose the raw OS handle. I already have to do some P/Invoke work for other reasons (setting up security on the pipe is outside the abilities of the runtime class so I use P/Invoke to create it). I just need to add the code that sends a zero byte payload using P/Invoke for just the end piece. It's solvable, just a little messy.

**Reply by mconnew** [reply]
The bigger issue if this is being installed on custom sites is the need to configure Kestrel. You don't want it listening on a random http port as that might flag security stuff. Instead configure Kestrel to listen on either a named pipe which will be unused, or a unix domain socket. That way it won't trigger any firewall UI stuff.

**Reply by cbenard** [reply]
Side question. Prior to making this issue and trying Unix Domain Sockets, I tried to roll my own poor-man's WCF with DataContractSerializer (gzip compressed) using `NamedPipeServerStream` and `NamedPipeClientStream`. I didn't try to multiplex anything over the same connection, since that would have greatly increased complexity with concurrency of requests, tracking request IDs, some kind of queue, etc.

So, I tried to treat it like I would a TCP `Socket` and do the `BeginWaitForConnection` (using async pattern in old code), immediately spinning up another `BeginWaitForConnection` before doing `EndWaitForConnection` (like I'd do with `BeginAcceptClient` or whatever).

The problem I ran into was that I seemed to be limited to 2-3 concurrent clients hitting the same named pipe (different `NamedPipeServerStream` instances). I couldn't find anything about that limitation in Windows, and I'm passing `NamedPipeServerStream.MaxAllowedServerInstances` and mode is `Asychronous`.

The server just wouldn't see the clients after 2-3 connecting until one of the first 2-3 disconnected. I was limited to about 2-3 concurrently.

At that point, I gave up, because I'm assuming WCF must be multiplexing the requests, similar to HTTP/2. I'd like to stay in the WCF world anyway, since that's what the .NET Framework code is using, but I was trying to explore alternatives, since we don't have `NetNamedPipeBinding` in `CoreWCF` yet.

At that point, I started doing a POC with the Unix Domain Sockets, since I didn't realize `AF_UNIX` had been added, but I recently found out we have Win10 clients still in use, and I need to be able to support them.

That leaves me probably needing to use named pipes, or downgrade to localhost TCP on some "reserved" port, but I expect to have problems with some client firewall software if we go that route. I'd really like to stick with named pipes.

**Reply by cbenard** [reply]
> Instead configure Kestrel to listen on either a named pipe which will be unused, or a unix domain socket.

The actual code I'm using the following. Is this enough to not start up an HTTP(S) listener? I only intend to be running a socket listener. The debug output doesn't seem to show it listening on a TCP port.

I redacted a couple of things (Serilog sink and service names)

```cs
WebApplicationBuilder hostBuilder = WebApplication.CreateSlimBuilder();
hostBuilder.Host.UseUnixDomainSocket(options =>
{
    options.Listen(new Uri(TheServiceOptions.UnixSocketClientAddress));
});
hostBuilder.Services.AddServiceModelServices();
hostBuilder.Services.AddSingleton<TheService>(this);

hostBuilder.Services.AddLogging();
hostBuilder.Host.UseSerilog((ctx, lc) => lc
    .MinimumLevel.Warning()
    .WriteTo.OtherSink());

IHost app = hostBuilder.Build();

CoreWCF.UnixDomainSocketBinding serverBinding = new CoreWCF.UnixDomainSocketBinding(CoreWCF.UnixDomainSocketSecurityMode.None);

app.UseServiceModel(builder =>
{
    builder.AddService<TheService>(options => options.DebugBehavior.IncludeExceptionDetailInFaults = true);
    builder.AddServiceEndpoint<TheService, ITheService>(serverBinding, TheServiceOptions.UnixSocketClientAddress);
});

await app.StartAsync();
await app.WaitForShutdownAsync(_shutdownCts.Token);
```

**Reply by cbenard** [reply]
Again, as a side note, I cannot use the default constructor on `UnixDomainSocketBinding`. It fails with an exception which lead me down the rabbit hole of trying to tweak security/protection options.

Please see my latest stack trace as a reply to you in #1411.

**Reply by mconnew** [reply]
If you use NetTcp, and specify IPAddress.Loopback, I don't think it will trigger any firewall stuff. But I totally understand not wanting to deal with a customer using a 3rd party firewall product which happens to trigger on localhost listeners. Too many unknowns to switch over if you already have a large enough customer base.

**Reply by cbenard** [reply]
Hi @mconnew. I am circling back to your comments before. You mentioned:

> And there's a problem where it requires configuring kestrel to listen on a benign port.

Is that still true? The last comments on #1336 seemed to indicate there was a resolution to that, but I'm not sure if you're referring to a different issue.

We discussed the exception handling, so I think that's good. And, like I said, we don't need WSDL. Do you see any reason we can't use the `-release1` version of the named pipe binding with the actual released version of CoreWCF?

I assume you don't want to release it without WSDL, but I don't understand the "benign port" issue at this point.

Thank you again.
